File: PPD/app/src/main/java/com/yassou/postpartumapp/fragments/RadioBoxesFragment.java
package com.yassou.postpartumapp.fragments;

import android.os.Bundle;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;

import com.yassou.postpartumapp.Activities.QuestionaireActivity;
import com.yassou.postpartumapp.R;
import com.yassou.postpartumapp.model.ChoicesDataModel;
import com.yassou.postpartumapp.model.QuestionsDataModel;

import java.util.ArrayList;
import java.util.List;

/**
 * This fragment provide the RadioButton/Single Options.
 */
public class RadioBoxesFragment extends Fragment {
    private final ArrayList<RadioButton> radioButtonArrayList = new ArrayList<>();

    private FragmentActivity mContext;
    private Button nextOrFinishButton;
    private QuestionsDataModel radioButtonTypeQuestion;
    private TextView questionRBTypeTextView;
    private RadioGroup radioGroupForChoices;
    private boolean atLeastOneChecked = false;
    private int currentPagePosition = 0;
    private int clickedRadioButtonPosition = 0;
    int score;
    String scale;

    public RadioBoxesFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        Bundle bundle = getArguments();

        if (getArguments() != null) {
            radioButtonTypeQuestion = (QuestionsDataModel) bundle.getSerializable("question");
            currentPagePosition = getArguments().getInt("page_position") + 1;
        }

        ViewGroup rootView = (ViewGroup) inflater.inflate(R.layout.fragment_radio_boxes, container, false);

        nextOrFinishButton = rootView.findViewById(R.id.nextOrFinishButton);
        questionRBTypeTextView = rootView.findViewById(R.id.questionRBTypeTextView);
        questionRBTypeTextView.setText(radioButtonTypeQuestion.getQuestion());
        radioGroupForChoices = rootView.findViewById(R.id.radioGroupForChoices);


        nextOrFinishButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (currentPagePosition == ((QuestionaireActivity) mContext).questionaireTotalQuestions()) {
                    ((QuestionaireActivity) mContext).goToResultsActivity();
                } else {
                    ((QuestionaireActivity) mContext).nextQuestion(score) ;
                }

            }
        });


        return rootView;
    }

    private void initialQuestionSetup() {
        ((QuestionaireActivity) mContext).updateQuestionnaireProgress(0);

    }

    private void saveActionsOfRadioBox() {
        for (int i = 0; i < radioButtonArrayList.size(); i++) {
            if (i == clickedRadioButtonPosition) {
                RadioButton radioButton = radioButtonArrayList.get(i);
                if (radioButton.isChecked()) {
                    atLeastOneChecked = true;

                }
            }
        }

        if (atLeastOneChecked) {
            nextOrFinishButton.setEnabled(true);
        } else {
            nextOrFinishButton.setEnabled(false);
        }
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);
    }


    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mContext = getActivity();


        List<ChoicesDataModel> choices = radioButtonTypeQuestion.getChoices();
        radioButtonArrayList.clear();

        for (final ChoicesDataModel choice : choices) {
            RadioButton rb = new RadioButton(mContext);
            rb.setText(choice.getTitle());
            rb.setTextSize(TypedValue.COMPLEX_UNIT_SP, 18);
            rb.setTextColor(ContextCompat.getColor(mContext, R.color.purple));
            rb.setPadding(10, 40, 10, 40);
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
            params.leftMargin = 25;
            rb.setLayoutParams(params);

            View view = new View(mContext);
            view.setBackgroundColor(ContextCompat.getColor(mContext, R.color.orchid));
            view.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 1));

            radioGroupForChoices.addView(rb);
            radioGroupForChoices.addView(view);
            radioButtonArrayList.add(rb);

            rb.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                    clickedRadioButtonPosition = radioButtonArrayList.indexOf(buttonView);
                    RadioBoxesFragment.this.saveActionsOfRadioBox();
                    if (isChecked) {
                        score = choice.getValue();
                    }
                }
            });

        }

        if (atLeastOneChecked) {
            nextOrFinishButton.setEnabled(true);
        } else {
            nextOrFinishButton.setEnabled(false);
        }

        /* If the current question is last in the questionnaire then
        the "Next" button will change into "Finish" button*/
        if (currentPagePosition == ((QuestionaireActivity) mContext).questionaireTotalQuestions()) {
            nextOrFinishButton.setText(R.string.finish);
        } else {
            nextOrFinishButton.setText(R.string.next);
        }
        initialQuestionSetup();
    }
}

File: PPD/app/src/main/java/com/yassou/postpartumapp/fragments/RadioBoxesFragment.kt
package com.yassou.postpartumapp.fragments

import android.os.Build
import android.os.Bundle
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.annotation.RequiresApi
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.yassou.postpartumapp.Activities.QuestionaireActivity
import com.yassou.postpartumapp.R
import com.yassou.postpartumapp.model.QuestionsDataModel
import java.util.*


/**
 * This fragment provide the RadioButton/Single Options.
 */
class RadioBoxesFragment : Fragment() {
    private val radioButtonArrayList = ArrayList<RadioButton>()

    private var mContext: FragmentActivity? = null
    private var nextOrFinishButton: Button? = null
    private var radioButtonTypeQuestion: QuestionsDataModel? = null
    private var questionRBTypeTextView: TextView? = null
    private var radioGroupForChoices: RadioGroup? = null
    private var atLeastOneChecked = false
    private var currentPagePosition = 0
    private var clickedRadioButtonPosition = 0
    internal var score: Int = 0

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val bundle = arguments

        if (arguments != null) {
            radioButtonTypeQuestion = bundle!!.getSerializable("question") as QuestionsDataModel
            currentPagePosition = arguments!!.getInt("page_position") + 1
        }

        val rootView =
            inflater.inflate(R.layout.fragment_radio_boxes, container, false) as ViewGroup

        nextOrFinishButton = rootView.findViewById(R.id.nextOrFinishButton)
        questionRBTypeTextView = rootView.findViewById(R.id.questionRBTypeTextView)
        questionRBTypeTextView!!.text = radioButtonTypeQuestion!!.question
        radioGroupForChoices = rootView.findViewById(R.id.radioGroupForChoices)


        nextOrFinishButton!!.setOnClickListener {
            if (currentPagePosition == (mContext as QuestionaireActivity).questionaireTotalQuestions()) {
                (mContext as QuestionaireActivity).goToResultsActivity()
            } else {
                (mContext as QuestionaireActivity).nextQuestion(score)
            }
        }


        return rootView
    }

    private fun initialQuestionSetup() {
        (mContext as QuestionaireActivity).updateQuestionnaireProgress(0)

    }

    private fun saveActionsOfRadioBox() {
        for (i in radioButtonArrayList.indices) {
            if (i == clickedRadioButtonPosition) {
                val radioButton = radioButtonArrayList[i]
                if (radioButton.isChecked) {
                    atLeastOneChecked = true

                }
            }
        }

        if (atLeastOneChecked) {
            nextOrFinishButton!!.isEnabled = true
        } else {
            nextOrFinishButton!!.isEnabled = false
        }
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
    }


    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        mContext = activity


        val choices = radioButtonTypeQuestion!!.choices
        radioButtonArrayList.clear()

        for (choice in choices!!) {
            val rb = RadioButton(mContext)
            rb.text = choice.title
            rb.setTextSize(TypedValue.COMPLEX_UNIT_SP, 18f)
            rb.setTextColor(ContextCompat.getColor(mContext!!, R.color.black))
            rb.setPadding(10, 40, 10, 40)

            rb.setBackgroundColor(ContextCompat.getColor(mContext!!,R.color.colorBackground))

            val params = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            params.leftMargin = 25
            rb.layoutParams = params

            val view = View(mContext)
            view.setBackgroundColor(ContextCompat.getColor(mContext!!, R.color.colorBackground))
            view.layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 1)

            radioGroupForChoices!!.addView(rb)
            radioGroupForChoices!!.addView(view)
            radioButtonArrayList.add(rb)

            rb.setOnCheckedChangeListener { buttonView, isChecked ->
                clickedRadioButtonPosition = radioButtonArrayList.indexOf(buttonView)
                this@RadioBoxesFragment.saveActionsOfRadioBox()
                if (isChecked) {
                    score = choice.value!!
                }
            }

        }

        if (atLeastOneChecked) {
            nextOrFinishButton!!.isEnabled = true
        } else {
            nextOrFinishButton!!.isEnabled = false
        }

        /* If the current question is last in the questionnaire then
        the "Next" button will change into "Finish" button*/
        if (currentPagePosition == (mContext as QuestionaireActivity).questionaireTotalQuestions()) {
            nextOrFinishButton!!.setText(R.string.finish)
        } else {
            nextOrFinishButton!!.setText(R.string.next)
        }
        initialQuestionSetup()
    }
}